README.md
# Svelte Piano

A typescript / svelte rewrite of the awesome [AudioKeys](https://github.com/kylestetz/AudioKeys) project.

From the original project:

> Use AudioKeys to power the QWERTY keyboard in your next Web Audio project. AudioKeys provides **intelligent handling of key events**, giving you key up and key down events that you can use to trigger your sounds.
>
> AudioKeys provides **configurable polyphony**— if you're making a monophonic synth, choose from the common note priorities "last note", "first note", "highest note", or "lowest note". It also handles odd situations like switching tabs— AudioKeys fires a note off event when your browser window goes out of focus.

# Installation

### Using `pnpm`

```bash
pnpm install -D svelte-piano
```


# Usage

```html
<script>
	import { Piano } from 'svelte-piano';
</script>

<Piano />
```

<br>

## Options


|             |                                                             |
|-------------|-------------------------------------------------------------|
| `polyphony` | The number of keys that can be active simultaneously.       |
| `notes`     | Whether or not to display the note names.  Default `false`. |
| `sound`     | Whether to load and play the stock keyboard sound.          |
| `width`     | Piano width in pixels.                                      |
| `height`    | Piano height in pixels.                                     |
| `theme`     | `todo`                                                      |

```jsx
<Piano
	options={
		polyphony: 1,
		notes: false,
		keys: true,
		sound: true
	}
	'--width': 900px;
	'--height': 250px;
 />
```

<br>

## Note events

You can subscribe to note events using the `onKeyDown`, `onKeyUp`, and `activeKeys` stores.

```html
<script>
	import { SveltePiano, onKeyDown, onKeyUp, activeKeys } from 'svelte-piano';

	$: console.log($activeKeys)

	$: $onKeyDown, () => {
		console.log($onKeyDown)
	}

	$: $onKeyUp, () => {
		console.log($onKeyUp)
	}
</script>

<Piano />
```

### The `Note` interface:

```typescript
interface Note {
	/**
	 * The keyCode of the key being pressed down.
	 */
	keyCode: number

	/**
	 * The midi number of the note.
	 */
	note: number

	/**
	 * The frequency of the note between 0 and 20,000.
	 */
	frequency: number

	/**
	 * On note down, the current velocity.
	 * On note up, 0.
	 */
	velocity: number

	/**
	 * Whether the key is currently being pressed down.
	 */
	isActive: boolean
}
```

These properties will be useful in setting up instruments. See the [`lib/Instrument.ts`](https://github.com/fractalhq/svelte-piano/blob/main/src/lib/Instrument.ts) file for a simple example.

<br>

## Headless API

If you're not using Svelte, you can use the keyboard directly. Options can be passed into the `QwertyKeyboard` constructor in an object or set individually using `set`.

```ts
// Headless vanilla keyboard.
const keyboard = new QwertyKeyboard({
  polyphony: 1,
  rows: 2,
  priority: 'lowest'
});

// Properties can also be set later.
keyboard.priority = 'highest'
```

| Property           | Description                                                                                                                                                                                            |
|--------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `polyphony`        | The number of keys that can be active simultaneously.                                                                                                                                                  |
| `rows`             | Either `1` or `2`.                                                                                                                                                                                     |
| `octaveControls`   | Determines whether or not the `z` and `x` keys shift octaves when `rows` is set to `1`.                                                                                                                |
| `velocityControls` | Determines whether or not the number keys set the velocity of the notes being triggered.                                                                                                               |
| `priority`         | Determines the priority of the note triggers.                                                                                                                                                          |
|                    | `"last"`: prefer the last note(s) pressed. <br> `"first"`: prefer the first note(s) pressed. <br> `"highest"`: prefer the highest note(s) pressed. <br> `"lowest"`: prefer the lowest note(s) pressed. |
| `rootNote`         | Determines what note the lowest key on the keyboard will represent. The default is `60` (C4).                                                                                                          |

The default `rootNote` is `60` (C4). Keep in mind that setting it to a note other than C (36, 48, 60, 72, 84, etc.) will result in the key mappings not lining up like a regular keyboard!

For more on note priority, check out [this Sound on Sound article](https://web.archive.org/web/20150913012148/http://www.soundonsound.com/sos/oct00/articles/synthsec.htm).



CHANGELOG.md
# svelte-piano

## 0.1.1

### Patch Changes

- [fix] dep resolution issues ([`e1fc499`](https://github.com/FractalHQ/svelte-piano/commit/e1fc4992ae329ede9acbc610868a5d5c1a052a6e))

## 0.1.0

### Minor Changes

- [release] initial release ([`29045af`](https://github.com/FractalHQ/svelte-piano/commit/29045af330f725d959902fdf5adcd6978316e461))


src/lib/types/Options.ts
import type { Keymap, SpecialKeymap } from './Keymap'
import type { Note } from './Note'

export interface KeyboardOptions {
	/**
	 * The number of keys that can be active simultaneously.
	 * @default 10
	 */
	polyphony: number

	/**
	 * The number of rows of keys.
	 * @default 1
	 */
	rows: 1 | 2

	/**
	 * Determines the priority of the note triggers. Priority only takes effect when
	 * the number of keys being pressed down exceeds the polyphony (e.g. when the polyphony
	 * is 1 but a second key is pressed).
	 * - "last": prefer the last note(s) pressed
	 * - "first": prefer the first note(s) pressed
	 * - "highest": prefer the highest note(s) pressed
	 * - "lowest": prefer the lowest note(s) pressed
	 * For more on note priority, check out this {@link https://web.archive.org/web/20150913012148/http://www.soundonsound.com/sos/oct00/articles/synthsec.htm Sound on Sound article}.
	 * @default "last"
	 */
	priority: 'last' | 'first' | 'highest' | 'lowest'

	/**
	 * Determines what note the lowest key on the keyboard will represent.
	 * @default 60
	 * @remarks Keep in mind that setting it to a note other than `C` (`36`, `48`, `60`, `72`, `84`, etc.)
	 * will result in the key mappings not lining up like a regular keyboard!
	 */
	rootNote: number

	/**
	 * Determines whether or not the Z and X keys change the octaves when rows is set to 1.
	 */
	octaveControls: boolean

	/**
	 * The current / default octave.
	 */
	octave: number

	/**
	 * Determines whether or not the number keys set the velocity of the notes being triggered.
	 * Keep in mind that velocity is just a number— you have to interpret it in your sounds!
	 * @default true
	 */
	velocityControls: boolean

	/**
	 * Optional default note intensity.
	 */
	velocity: number

	/**
	 * A map of keyboard keys to midi notes.
	 */
	keymap?: Keymap

	/**
	 * A map of keyboard number keys to velocity amounts. `[1, 2 ... 9, 0]` maps to `[0 ... 127]`
	 */
	specialKeymap?: SpecialKeymap

	/**
	 * The width of the keyboard.
	 * @default 600
	 */
	width: number

	/**
	 * The height of the keyboard.
	 * @default 200
	 */
	height: number

	/**
	 * Whether or not event listeners should be added to the windows keyboard events automatically.
	 * @default true
	 **/
	eventListeners: boolean
}

export interface State extends KeyboardOptions {
	/**
	 * Array of notes currently being pressed down.
	 */
	keys: Note[]

	/**
	 * Array of notes currently being played
	 */
	buffer: Note[]
}
